A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,516 @@
+the `override-host` for the RPC. This value should be a syntactically valid
+IP:port address: if not, then log a local warning and pass the RPC through.
+For a valid `override-host` value, set it in the RPC context as the value of
+`upstreamAddress` state variable which will be used in the response processing
+of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path - if the filter was not skipped based on the RPC
+  path - get the value of `upstreamAddress`. Also get the peer address for the
+  RPC; let’s call it `hostAddress` here. For example, in Java you get the peer
+  address through the `Grpc.TRANSPORT_ATTR_REMOTE_ADDR` attribute of the
+  current ClientCall. If `upstreamAddress` is not set or `upstreamAddress` and
+  `hostAddress` are different, then create a cookie using our Cookie config
+  which has the `name`, `ttl` and `path` values. Set the value of the cookie
+  to be the base64-encoded string value of `hostAddress`. Add this Cookie to
+  the response headers. The logic in pseudo-code is:
+
+```C++
+  if (!upstreamAddress.has_value() || hostAddress != upstreamAddress) {
+    String encoded_address = base64_encode(hostAddress);
+    Create a Cookie from our cookie configuration;
+    Set cookie's value as encoded_address;
+    Add the cookie header to the response;
+  }
+```
+
+### LB Policy for Stateful Session Affinity
+
+After the `CookieBasedStatefulSessionFilter` has passed the `override-host`
+value to the Load Balancer as a "pick-argument", the Load Balancer uses
+this value to route the RPC appropriately. The required logic will be
+implemented in a new and separate LB policy at an appropriate place in the
+hierarchy.
+
+RPC load reporting happens in the `xds_cluster_impl_experimental` policy and
+we do want all RPCs to be included in the load reports. Hence the new policy
+needs to be just below the `xds_cluster_impl_experimental` as its child policy.
+Note that the new policy will be beneath the `priority_experimental` policy,
+which means that if we pick a host for a session while we are failed over to a
+lower priority and then a higher priority comes back online, we will switch
+hosts at that point and this behavior is different from Envoy.
+
+Let's call the new policy `xds_override_host_experimental`. This policy
+contains subchannel management, endpoint management, and RPC routing logic as
+follows:
+
+* maintain a map - let's call it `address_map`. The key is (IP:port)
+  and the value is the tuple
+  `[subchannel, EDS health-status, list of other equivalent addresses]`.
```

Comment:
The map will also need to store the current connectivity state of the subchannel.

Also, I think that we will need to wrap all of the subchannels in order to get connectivity state updates for them.

Reply:
done